src/scaraplate/automation/git.py

```diff
@@ -0,0 +1,356 @@
+
+        :param clone_url: Any valid ``git clone`` url.
+        :param clone_ref: Git ref to checkout after clone
+            (i.e. branch or tag name).
+        :param monorepo_inner_path: Path to the root dir of project
+            relative to the root of the repo. If None, the root of
+            the repo will be used as the root of project.
+        :param changes_branch: The branch name where the changes should be
+            pushed in the remote. Might be the same as ``clone_ref``.
+            Note that this branch is never force-pushed. If upon push
+            the branch already exists in remote and its one-commit diff
+            is different from the one-commit diff of the just created
+            local branch, then the remote branch will be deleted and
+            the local branch will be pushed to replace the previous one.
+        :param commit_author: Author name to use for ``git commit``, e.g.
+            ``John Doe <john@example.org>``.
+        :param commit_message_template: :meth:`str.format` template
+            which is used to produce a commit message when committing
+            the changes. Available format variables are:
+
+            - ``update_time`` [:class:`datetime.datetime`] -- the time
+              of update
+            - ``scaraplate_version`` [:class:`str`] -- scaraplate package
+              version
+            - ``template_meta`` [:class:`.TemplateMeta`] -- template meta
+              returned by :meth:`.TemplateVCS.template_meta`
+        """
+
+        with tempfile.TemporaryDirectory() as tmpdir_name:
+            tmpdir_path = Path(tmpdir_name).resolve()
+            project_path = tmpdir_path / "scaraplate_project"
+
+            git = Git.clone(clone_url, target_path=project_path, ref=clone_ref)
+
+            if monorepo_inner_path is not None:
+                project_path = project_path / monorepo_inner_path
+
+            yield cls(
+                project_path,
+                git,
+                changes_branch=changes_branch,
+                commit_author=commit_author,
+                commit_message_template=commit_message_template,
+            )
+
+
+class Git:
+    def __init__(self, cwd: Path, remote: str = "origin") -> None:
+        self.cwd = cwd
+        self.remote = remote
+
+    def remote_ref(self, ref: str) -> str:
+        return f"{self.remote}/{ref}"
+
+    def checkout_branch(self, branch: str) -> None:
+        self._git(["checkout", "-b", branch])
+
+    def commit_all(self, commit_message: str, author: Optional[str] = None) -> None:
+        self._git(["add", "--all"])
+        extra: List[str] = []
+        if author is not None:
+            extra = ["--author", author]
+        self._git(
+            ["commit", "-m", commit_message, *extra],
+            env={
+                # git would fail if there's no `user.email` in the local
+                # git config, even if `--author` is specified.
+                "USERNAME": "scaraplate",
```

Comment:
These values are only required for creating a commit. The `commit_all` method already accepts an `author` argument which overrides these. It is mandatory in `GitCloneProjectVCS` and optional in `GitLabMRProjectVCS` with fallback to the authenticated gitlab user name and email.

So basically these values are never used in the resulting commits. They are here just to avoid git failing when `--author` is specified though no `user` is present in the local git config.

Reply:
Cool. If this is not an issue, let's propagate invalid git configuration back to user? Yes, that's not friendly, but no more like add unknown guy to git history. I guess, in our production this is a problem and we just have to resolve it anyway.

Reply:
The only thing that needs to be configured in git is author name, and it is already a required argument in the automation classes. So I don't see any reason to require anyone to *also* fill *something* to their git config, making automation setup more complicated.

A missing user in the local git config is not a misconfiguration in this case.

Reply:
I guess, if someone starts to automate his processes it should know how things are works anyway. Compromise option is to keep these defaults and if they get applied log a warning about their usage. And everyone should be good then.